src/main/java/com/timgroup/statsd/StatsDProcessor.java

```diff
@@ -0,0 +1,128 @@
+    private final Queue<String> messages;
+    private final AtomicInteger qSize;  // qSize will not reflect actual size, but a close estimate.
+    private final BlockingQueue<ByteBuffer> outboundQueue; // FIFO queue with outbound buffers
+
+    private volatile boolean shutdown;
+
+
+    StatsDProcessor(final int queueSize, final StatsDClientErrorHandler handler,
+            final int maxPacketSizeBytes, final int poolSize)
+            throws Exception {
+
+        this.bufferPool = new BufferPool(poolSize, maxPacketSizeBytes, true);
+
+        this.qSize = new AtomicInteger(0);
+        this.qCapacity = queueSize;
+
+        this.messages = new ConcurrentLinkedQueue<String>();
+        this.outboundQueue = new ArrayBlockingQueue<ByteBuffer>(poolSize);
+
+        this.handler = handler;
+    }
+
+    boolean send(final String message) {
+        if (!shutdown) {
+            if (qSize.get() < qCapacity) {
+                messages.offer(message);
+                qSize.incrementAndGet();
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    public BufferPool getBufferPool() {
+        return this.bufferPool;
+    }
+
+    public BlockingQueue<ByteBuffer> getOutboundQueue() {
+        return this.outboundQueue;
+    }
+
+    @Override
+    public void run() {
+        boolean empty;
+        ByteBuffer sendBuffer;
+
+        try {
+            sendBuffer = bufferPool.borrow();
+        } catch(final InterruptedException e) {
+            handler.handle(e);
+            return;
+        }
+
+        while (!((empty = messages.isEmpty()) && shutdown)) {
+
+            try {
+                if (empty) {
+                    Thread.sleep(WAIT_SLEEP_MS);
+                    continue;
+                }
+
+                if (Thread.interrupted()) {
+                    return;
+                }
+                final String message = messages.poll();
+                if (message != null) {
+                    qSize.decrementAndGet();
+                    final byte[] data = message.getBytes(MESSAGE_CHARSET);
+                    if (sendBuffer.capacity() < data.length) {
+                        throw new InvalidMessageException(MESSAGE_TOO_LONG, message);
+                    }
+                    if (sendBuffer.remaining() < (data.length + 1)) {
+                        outboundQueue.put(sendBuffer);
+                        sendBuffer = bufferPool.borrow();
+                    }
+                    if (sendBuffer.position() > 0) {
+                        sendBuffer.put((byte) '\n');
+                    }
+                    sendBuffer.put(data);
+                    if (null == messages.peek()) {
+                        outboundQueue.put(sendBuffer);
+                        sendBuffer = bufferPool.borrow();
+                    }
+                }
+            } catch (final InterruptedException e) {
+                if (shutdown) {
+                    return;
```

Comment:
Do we also want to call `handler.handle(e);` as in line 76?

Reply:
It's a good question, I'm not sure, we're going to gracefully shutdown, and handle doesn't do much, but it's a good question. 🤔

Reply:
Let me think about it, I'll address in #94 if I decide we should handle these errors differently.